File: research/protocol/walkplay.py
"""WalkPlay (SchemeNo11) HID PEQ helpers for CrinEar Protocol Micro (3302:C20F).

Mirrors devicePEQ walkplayHidHandler.js @ 0617f382e76629792a5933e6933e4b396a756a93.
hidapi buffers include the report ID at [0]; WebHID `data[k]` == buf[k+1].
"""
import math
import time

import hid

VID, PID = 0x3302, 0xC20F
REPORT_ID = 0x4B
REPORT_LEN = 63            # payload bytes after report ID (HID descriptor: 0x4B, count 0x3F)
READ, WRITE, END = 0x80, 0x01, 0x00
CMD_GLOBAL_GAIN, CMD_PEQ, CMD_VERSION = 0x03, 0x09, 0x0C
MAX_FILTERS = 8
FREQ_FACTOR = 0.9775       # freqCompensation ratio (SchemeNo11)
DESIGN_FS = 96000          # qCompensation cosNyquist designFs
TYPE_NAMES = {1: "LSQ", 2: "PK", 3: "HSQ", 4: "LP", 5: "HP"}
TYPE_CODES = {v: k for k, v in TYPE_NAMES.items()}
APO_TYPES = {"PK": "PK", "LSQ": "LSC", "HSQ": "HSC", "LP": "LPQ", "HP": "HPQ"}


def hx(b):
    return " ".join(f"{x:02x}" for x in b)


class Log:
    def __init__(self, path=None, echo=True):
        self.f = open(path, "a", encoding="utf-8") if path else None
        self.echo = echo
        self.t0 = time.perf_counter()

    def __call__(self, direction, data):
        line = f"{(time.perf_counter() - self.t0) * 1000:9.1f} ms {direction} {hx(data)}"
        if self.echo:
            print(line)
        if self.f:
            self.f.write(line + "\n")
            self.f.flush()


def find_path():
    devs = [d for d in hid.enumerate(VID, PID) if d["interface_number"] == 3]
    if not devs:
        raise SystemExit("Protocol Micro HID interface (MI_03) not found")
    return devs[0]["path"]


class Device:
    def __init__(self, log, dry_run=False):
        self.log, self.dry = log, dry_run
        self.h = None
        if not dry_run:
            self.h = hid.device()
            self.h.open_path(find_path())

    def close(self):
        if self.h:
            self.h.close()

    def send(self, payload):
        buf = bytes([REPORT_ID]) + bytes(payload) + bytes(REPORT_LEN - len(payload))
        self.log("TX" + (" (dry-run, not sent)" if self.dry else ""), buf)
        if not self.dry:
            n = self.h.write(buf)
            if n < 0:
                raise IOError(f"hid write failed: {self.h.error()}")

    def drain(self, ms):
        """Collect every 0x4B input report arriving within `ms`."""
        out = []
        if self.dry:
            time.sleep(ms / 1000)
            return out
        end = time.perf_counter() + ms / 1000
        while True:
            left = int((end - time.perf_counter()) * 1000)
            if left <= 0:
                break
            r = self.h.read(64, left)
            if r:
                self.log("RX", r)
                if r[0] == REPORT_ID:
                    out.append(bytes(r))
        return out

    def request(self, payload, cmd, timeout_ms=2000, attempts=3):
        """Send and wait for a reply whose data[1] (WebHID) == cmd."""
        for attempt in range(1, attempts + 1):
            self.send(payload)
            end = time.perf_counter() + timeout_ms / 1000
            while time.perf_counter() < end:
                r = self.h.read(64, max(1, int((end - time.perf_counter()) * 1000)))
                if r:
                    self.log("RX", r)
                    if r[0] == REPORT_ID and r[2] == cmd:
                        return bytes(r)
            print(f"  no reply to cmd 0x{cmd:02x} (attempt {attempt}/{attempts})")
        raise TimeoutError(f"device did not answer cmd 0x{cmd:02x} after {attempts} attempts")


def parse_filter(buf):
    """buf includes report ID. Offsets follow parseFilterPacket (+1)."""
    d = buf[1:]
    idx = d[4]
    freq_raw = d[27] | (d[28] << 8)
    q_raw = d[29] | (d[30] << 8)
    gain_raw = d[31] | (d[32] << 8)
    if gain_raw > 32767:
        gain_raw -= 65536
    type_code = d[33]
    typ = TYPE_NAMES.get(type_code, "PK")
    gain = round(gain_raw / 256, 2)
    q_stored = round(q_raw / 256, 2)
    freq = freq_raw * FREQ_FACTOR if freq_raw > 0 else freq_raw
    ratio = math.cos(math.pi * min(max(freq_raw / DESIGN_FS, 0), 0.5 - 1e-9)) if typ in ("PK", "LSQ", "HSQ") else 1.0
    q = q_stored * ratio if q_stored > 0 else q_stored
    uninit = freq_raw == 0xFFFF and q_raw == 0xFFFF and gain_raw == -1
    disabled = uninit or freq_raw in (0, 0xFFFF) or q == 0
    return {
        "index": idx,
        "enabled": not disabled,
        "type": typ,
        "freq_hz": round(freq, 1),
        "gain_db": gain,
        "q": round(q, 3),
        "raw": {"freq": freq_raw, "q_x256": q_raw, "gain_x256": gain_raw, "type": type_code,
                "slot_byte": d[35], "biquad_hex": hx(d[7:27])},
    }


def read_all(dev):
    """devicePEQ read path: getCurrentSlot() then pullFromDevice()."""
    t = {}
    t0 = time.perf_counter()
    v = dev.request([READ, CMD_VERSION, END], CMD_VERSION)
    version = bytes(v[4:7]).decode("ascii", "replace")          # response.slice(3,6)
    t["version_ms"] = (time.perf_counter() - t0) * 1000
    t1 = time.perf_counter()
    s = dev.request([READ, CMD_PEQ, END], CMD_PEQ)
    slot = s[36]                                                 # response[35]
    t["slot_ms"] = (time.perf_counter() - t1) * 1000

    filters, per_band = {}, {}
    t2 = time.perf_counter()
    for i in range(MAX_FILTERS):
        ts = time.perf_counter()
        dev.send([READ, CMD_PEQ, 0x00, 0x00, i, END])
        for r in dev.drain(50):                                  # devicePEQ: delay(50) between requests
            if r[2] == CMD_PEQ and len(r) >= 33:
                f = parse_filter(r)
                filters[f["index"]] = f
                per_band.setdefault(f["index"], (time.perf_counter() - ts) * 1000)
    tries = 0
    while len(filters) < MAX_FILTERS and tries < 3:              # devicePEQ waits up to 10 s
        for r in dev.drain(500):
            if r[2] == CMD_PEQ and len(r) >= 33:
                f = parse_filter(r)
                filters[f["index"]] = f
        tries += 1
    t["bands_ms"] = (time.perf_counter() - t2) * 1000
    t3 = time.perf_counter()
    g = dev.request([READ, CMD_GLOBAL_GAIN, END], CMD_GLOBAL_GAIN, timeout_ms=1000)
    gg = g[5] - 256 if g[5] > 127 else g[5]                      # Int8(data[4])
    t["gain_ms"] = (time.perf_counter() - t3) * 1000
    t["total_ms"] = (time.perf_counter() - t0) * 1000
    t["per_band_first_reply_ms"] = {k: round(v, 1) for k, v in sorted(per_band.items())}
    missing = [i for i in range(MAX_FILTERS) if i not in filters]
    return {
        "device": "CrinEar Protocol Micro", "vid": "0x3302", "pid": "0xC20F",
        "firmware_version": version, "current_slot": slot,
        "preamp_db": gg,
        "filters": [filters[i] for i in sorted(filters)],
        "missing_filters": missing,
        "compensation": {"freq_factor": FREQ_FACTOR, "q_model": "cosNyquist", "design_fs": DESIGN_FS,
                         "note": "freq_hz/q are devicePEQ-decompensated (realised); raw.* are register values"},
        "timing_ms": {k: (round(v, 1) if isinstance(v, float) else v) for k, v in t.items()},
    }


def to_apo(state):
    lines = [f"Preamp: {state['preamp_db']:.1f} dB"]
    for n, f in enumerate(state["filters"], 1):
        on = "ON" if f["enabled"] else "OFF"
        typ = APO_TYPES[f["type"]]
        if f["type"] in ("LP", "HP"):
            lines.append(f"Filter {n}: {on} {typ} Fc {f['freq_hz']:.0f} Hz Q {f['q']:.2f}")
        else:
            lines.append(f"Filter {n}: {on} {typ} Fc {f['freq_hz']:.0f} Hz Gain {f['gain_db']:.1f} dB Q {f['q']:.2f}")
    return "\n".join(lines) + "\n"


def compute_iir(freq, gain, q, typ="PK"):
    """Port of computeIIRFilter (walkplayHidHandler.js L742-822): 5 x int32 LE, Q30."""
    A = math.sqrt(10 ** (gain / 20))
    w0 = freq * 6.283185307179586 / 96000
    s, c = math.sin(w0), math.cos(w0)
    alpha = s / (2 * q)
    if typ in ("LSQ", "HSQ"):
        sa = (s / 2) * math.sqrt((A + 1 / A) * (1 / q - 1) + 2)
        k = 2 * math.sqrt(A) * sa
        if typ == "LSQ":
            b0 = A * ((A + 1) - (A - 1) * c + k); b1 = 2 * A * ((A - 1) - (A + 1) * c)
            b2 = A * ((A + 1) - (A - 1) * c - k); a0 = (A + 1) + (A - 1) * c + k
            a1 = -2 * ((A - 1) + (A + 1) * c); a2 = (A + 1) + (A - 1) * c - k
        else:
            b0 = A * ((A + 1) + (A - 1) * c + k); b1 = -2 * A * ((A - 1) + (A + 1) * c)
            b2 = A * ((A + 1) + (A - 1) * c - k); a0 = (A + 1) - (A - 1) * c + k
            a1 = 2 * ((A - 1) - (A + 1) * c); a2 = (A + 1) - (A - 1) * c - k
    else:  # PK; note LP/HP also fall through to the PK formula in devicePEQ
        b0 = 1 + alpha * A; b1 = -2 * c; b2 = 1 - alpha * A
        a0 = 1 + alpha / A; a1 = -2 * c; a2 = 1 - alpha / A
    S = 1073741824
    vals = [round(b0 / a0 * S), round(b1 / a0 * S), round(b2 / a0 * S), -round(a1 / a0 * S), -round(a2 / a0 * S)]
    out = b""
    for v in vals:
        out += (v & 0xFFFFFFFF).to_bytes(4, "little")
    return out


def js_round(x):
    return math.floor(x + 0.5)


def filter_write_payload(index, freq_sent, gain, q_sent, typ, slot):
    """pushToDevice packet (L153-163). freq_sent/q_sent are ALREADY compensated."""
    biquad = compute_iir(freq_sent, gain, q_sent, typ)
    f = int(freq_sent) & 0xFFFF                      # JS (value >> 0) truncates
    qq = js_round(q_sent * 256) & 0xFFFF
    gg = js_round(gain * 256) & 0xFFFF
    return ([WRITE, CMD_PEQ, 0x18, 0x00, index, 0x00, 0x00] + list(biquad)
            + [f & 0xFF, f >> 8, qq & 0xFF, qq >> 8, gg & 0xFF, gg >> 8, TYPE_CODES[typ], 0x00, slot, END])


def global_gain_payload(db):
    return [WRITE, CMD_GLOBAL_GAIN, 0x02, 0x00, js_round(db) & 0xFF]


# devicePEQ commit sequence after filter writes (L182-188) - PERSISTS to flash.
COMMIT_SEQUENCE = [
    [WRITE, 0x05, END],
    [WRITE, 0x17, END],
    [WRITE, 0x0A, 0x04, 0x00, 0x00, 0xFF, 0xFF, END],   # TEMP_WRITE
    [WRITE, 0x01, 0x01, END],                            # FLASH_EQ enable+persist (slot byte absent)
]

File: android/core/src/main/kotlin/io/github/chronosauros/contour/core/WalkPlay.kt
package io.github.chronosauros.contour.core

import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * WalkPlay HID PEQ protocol, SchemeNo11 (CrinEar Protocol Micro, 3302:C20F).
 *
 * A 1:1 port of `research/protocol/walkplay.py`, itself a port of devicePEQ's walkplayHidHandler.js
 * (0BSD, Jerome O'Flaherty, commit 0617f382) verified bit-exact against the device.
 * Every buffer here is a full 64-byte report with the report ID at [0], so `buf[n] == d[n - 1]`
 * in PROTOCOL.md's WebHID offsets. JS semantics are kept: `Math.round` = floor(x + 0.5),
 * `>> 0` truncation, int32 wrap.
 */
object WalkPlay {
    const val VENDOR_ID: Int = 0x3302
    const val PRODUCT_ID: Int = 0xC20F
    const val REPORT_ID: Int = 0x4B
    const val REPORT_SIZE: Int = 64            // report ID + 63 payload bytes
    const val READ: Int = 0x80
    const val WRITE: Int = 0x01
    const val END: Int = 0x00
    const val CMD_GLOBAL_GAIN: Int = 0x03
    const val CMD_PEQ: Int = 0x09
    const val CMD_VERSION: Int = 0x0C
    const val BANDS: Int = 8
    const val FREQ_FACTOR: Double = 0.9775     // freqCompensation ratio (realised / requested)
    const val DESIGN_FS: Double = 96_000.0     // qCompensation cosNyquist designFs, and the biquad rate

    const val TYPE_LSQ: Int = 1
    const val TYPE_PK: Int = 2
    const val TYPE_HSQ: Int = 3
    const val TYPE_LP: Int = 4
    const val TYPE_HP: Int = 5

    /** What the device holds today in every slot (factory flat): PK, 0 dB, Q 0.75, these raw frequencies. */
    val FACTORY_FREQS: IntArray = intArrayOf(31, 62, 125, 250, 500, 1000, 2000, 4000)
    const val FACTORY_Q: Double = 0.75

    // ---- JS number semantics ---------------------------------------------------------------------

    /** JS `Math.round`. */
    fun jsRound(x: Double): Double = floor(x + 0.5)

    /** JS ToInt32 (`x >> 0`, `x & 0xFF` operands) for |x| < 2^63: truncate, then wrap modulo 2^32. */
    fun toInt32(x: Double): Int = x.toLong().toInt()

    fun typeCode(type: FilterType): Int = when (type) {
        FilterType.LOW_SHELF -> TYPE_LSQ
        FilterType.PEAK -> TYPE_PK
        FilterType.HIGH_SHELF -> TYPE_HSQ
        FilterType.LOW_PASS -> TYPE_LP
        FilterType.HIGH_PASS -> TYPE_HP
    }

    /** devicePEQ convertToFilterType: unknown codes read as PK. */
    fun filterType(code: Int): FilterType = when (code) {
        TYPE_LSQ -> FilterType.LOW_SHELF
        TYPE_HSQ -> FilterType.HIGH_SHELF
        TYPE_LP -> FilterType.LOW_PASS
        TYPE_HP -> FilterType.HIGH_PASS
        else -> FilterType.PEAK
    }

    private fun qCompensated(typeCode: Int): Boolean = typeCode == TYPE_PK || typeCode == TYPE_LSQ || typeCode == TYPE_HSQ

    // ---- reports ---------------------------------------------------------------------------------

    /** 64-byte output report: report ID, then [payload], zero padded. */
    fun report(vararg payload: Int): ByteArray {
        require(payload.size <= REPORT_SIZE - 1)
        val buf = ByteArray(REPORT_SIZE)
        buf[0] = REPORT_ID.toByte()
        payload.forEachIndexed { i, v -> buf[i + 1] = v.toByte() }
        return buf
    }

    fun versionRequest(): ByteArray = report(READ, CMD_VERSION, END)
    /** Bulk PEQ read (getCurrentSlot): the reply is band 0's packet, the slot byte at buf[36]. */
    fun slotRequest(): ByteArray = report(READ, CMD_PEQ, END)
    fun bandRequest(index: Int): ByteArray = report(READ, CMD_PEQ, 0x00, 0x00, index, END)
    fun preampRequest(): ByteArray = report(READ, CMD_GLOBAL_GAIN, END)

    private fun u8(buf: ByteArray, i: Int): Int = buf[i].toInt() and 0xFF
    private fun u16(buf: ByteArray, i: Int): Int = u8(buf, i) or (u8(buf, i + 1) shl 8)

    /** True for an input report 0x4B answering [cmd] (devicePEQ matches on d[1]). */
    fun isReply(buf: ByteArray, cmd: Int): Boolean =
        buf.size >= 3 && u8(buf, 0) == REPORT_ID && u8(buf, 2) == cmd

    /** Band reply for [index]; the bulk slot read answers with band 0 too. */
    fun isBandReply(buf: ByteArray, index: Int): Boolean =
        isReply(buf, CMD_PEQ) && buf.size >= 37 && u8(buf, 5) == index

    /** VERSION reply: ASCII at d[3..5], e.g. "0.2". */
    fun parseVersion(buf: ByteArray): String = String(buf.copyOfRange(4, 7), Charsets.US_ASCII).trimEnd('\u0000')

    fun parseSlot(buf: ByteArray): Int = u8(buf, 36)

    /** Global gain reply: int8 dB at d[4]. */
    fun parsePreamp(buf: ByteArray): Int = buf[5].toInt()

    /** Raw registers of one band, as the device stores them. [gain256] is signed. */
    data class Registers(val index: Int, val freq: Int, val q256: Int, val gain256: Int, val typeCode: Int)

    /** A decoded band reply: registers plus the devicePEQ-decompensated (realised) values. */
    data class DeviceBand(
        val registers: Registers,
        val slotByte: Int,
        val biquad: ByteArray,
        val enabled: Boolean,
        val type: FilterType,
        val freqHz: Double,
        val gainDb: Double,
        val q: Double,
    )

    /** parse_filter / parseFilterPacket. */
    fun parseBand(buf: ByteArray): DeviceBand {
        require(buf.size >= 37) { "band reply too short: ${buf.size}" }
        val index = u8(buf, 5)
        val freqRaw = u16(buf, 28)
        val qRaw = u16(buf, 30)
        var gainRaw = u16(buf, 32)
        if (gainRaw > 32767) gainRaw -= 65536
        val typeCode = u8(buf, 34)
        val type = filterType(typeCode)
        val gain = jsRound(gainRaw / 256.0 * 100) / 100
        val qStored = jsRound(qRaw / 256.0 * 100) / 100
        val freq = if (freqRaw > 0) decompensateFreq(freqRaw) else freqRaw.toDouble()
        val q = if (qStored > 0) decompensateQ(qStored, freqRaw, typeCode) else qStored
        val uninit = freqRaw == 0xFFFF && qRaw == 0xFFFF && gainRaw == -1
        val disabled = uninit || freqRaw == 0 || freqRaw == 0xFFFF || q == 0.0
        return DeviceBand(
            registers = Registers(index, freqRaw, qRaw, gainRaw, typeCode),
            slotByte = u8(buf, 36),
            biquad = buf.copyOfRange(8, 28),
            enabled = !disabled,
            type = type,
            freqHz = freq,
            gainDb = gain,
            q = q,
        )
    }

    // ---- compensation (SchemeNo11) -----------------------------------------------------------------

    /** Realised / requested Q at the frequency actually written ('cosNyquist', designFs 96 000). */
    fun qRatio(freqSent: Double, typeCode: Int): Double {
        if (!qCompensated(typeCode) || !(freqSent > 0)) return 1.0
        val x = minOf(maxOf(freqSent / DESIGN_FS, 0.0), 0.5 - 1e-9)
        return cos(Math.PI * x)
    }

    /** User frequency -> the frequency to write (unrounded; the biquad uses it as is). */
    fun compensateFreq(freqHz: Double): Double {
        if (!(freqHz > 0)) return freqHz
        return minOf(20_000.0, maxOf(20.0, freqHz / FREQ_FACTOR))
    }

    /** User Q -> the Q to write, keyed off the compensated frequency. */
    fun compensateQ(q: Double, freqSent: Double, typeCode: Int): Double {
        if (!(q > 0)) return q
        val ratio = qRatio(freqSent, typeCode)
        if (ratio == 1.0) return q
        return minOf(10.0, maxOf(0.1, q / ratio))
    }

    fun decompensateFreq(freqRaw: Int): Double = freqRaw * FREQ_FACTOR

    fun decompensateQ(qStored: Double, freqRaw: Int, typeCode: Int): Double = qStored * qRatio(freqRaw.toDouble(), typeCode)

    // ---- writes ------------------------------------------------------------------------------------

    /**
     * One band as it goes on the wire: [freq] and [q] are ALREADY compensated (or raw register values);
     * the biquad uses them unrounded, the metadata fields get `trunc(freq)`, `round(q * 256)`, `round(gain * 256)`.
     */
    data class BandWrite(val index: Int, val freq: Double, val gainDb: Double, val q: Double, val typeCode: Int) {
        /** The registers the device will hold after this write (for the read-back comparison). */
        fun registers(): Registers {
            var g = toInt32(jsRound(gainDb * 256)) and 0xFFFF
            if (g > 32767) g -= 65536
            return Registers(index, toInt32(freq) and 0xFFFF, toInt32(jsRound(q * 256)) and 0xFFFF, g, typeCode)
        }
    }

    /** Factory flat band of [index]: raw registers, no compensation - byte-identical to the device's own. */
    fun factoryFlat(index: Int): BandWrite = BandWrite(index, FACTORY_FREQS[index].toDouble(), 0.0, FACTORY_Q, TYPE_PK)

    /** A user band -> its wire values, through the SchemeNo11 compensation. */
    fun bandWrite(index: Int, band: Band): BandWrite {
        val code = typeCode(band.type)
        val f = compensateFreq(band.freqHz)
        return BandWrite(index, f, band.gainDb, compensateQ(band.q, f, code), code)
    }

    /** computeIIRFilter: 5 x int32 LE, Q30 - b0, b1, b2, -a1, -a2 normalised by a0, RBJ at 96 kHz. */
    fun computeIir(freq: Double, gainDb: Double, q: Double, typeCode: Int): ByteArray {
        val a = sqrt(10.0.pow(gainDb / 20))
        val w0 = freq * 6.283185307179586 / 96000
        val s = sin(w0)
        val c = cos(w0)
        val alpha = s / (2 * q)
        val b0: Double
        val b1: Double
        val b2: Double
        val a0: Double
        val a1: Double
        val a2: Double
        if (typeCode == TYPE_LSQ || typeCode == TYPE_HSQ) {
            val sa = (s / 2) * sqrt((a + 1 / a) * (1 / q - 1) + 2)
            val k = 2 * sqrt(a) * sa
            if (typeCode == TYPE_LSQ) {
                b0 = a * ((a + 1) - (a - 1) * c + k); b1 = 2 * a * ((a - 1) - (a + 1) * c)
                b2 = a * ((a + 1) - (a - 1) * c - k); a0 = (a + 1) + (a - 1) * c + k
                a1 = -2 * ((a - 1) + (a + 1) * c); a2 = (a + 1) + (a - 1) * c - k
            } else {
                b0 = a * ((a + 1) + (a - 1) * c + k); b1 = -2 * a * ((a - 1) + (a + 1) * c)
                b2 = a * ((a + 1) + (a - 1) * c - k); a0 = (a + 1) - (a - 1) * c + k
                a1 = 2 * ((a - 1) - (a + 1) * c); a2 = (a + 1) - (a - 1) * c - k
            }
        } else { // PK; devicePEQ uses the PK formula for LP/HP as well
            b0 = 1 + alpha * a; b1 = -2 * c; b2 = 1 - alpha * a
            a0 = 1 + alpha / a; a1 = -2 * c; a2 = 1 - alpha / a
        }
        val scale = 1073741824.0
        val values = doubleArrayOf(
            jsRound(b0 / a0 * scale), jsRound(b1 / a0 * scale), jsRound(b2 / a0 * scale),
            -jsRound(a1 / a0 * scale), -jsRound(a2 / a0 * scale),
        )
        val out = ByteArray(20)
        values.forEachIndexed { i, v ->
            val n = toInt32(v)
            out[i * 4] = (n and 0xFF).toByte()
            out[i * 4 + 1] = ((n shr 8) and 0xFF).toByte()
            out[i * 4 + 2] = ((n shr 16) and 0xFF).toByte()
            out[i * 4 + 3] = ((n shr 24) and 0xFF).toByte()
        }
        return out
    }

    /** Filter write `01 09 18 00 i 00 00` + biquad + freq + q + gain + type + 00 + slot + 00. */
    fun bandWriteReport(w: BandWrite, slot: Int): ByteArray {
        val biquad = computeIir(w.freq, w.gainDb, w.q, w.typeCode)
        val f = toInt32(w.freq) and 0xFFFF
        val q = toInt32(jsRound(w.q * 256)) and 0xFFFF
        val g = toInt32(jsRound(w.gainDb * 256)) and 0xFFFF
        val payload = IntArray(7 + 20 + 10)
        intArrayOf(WRITE, CMD_PEQ, 0x18, 0x00, w.index, 0x00, 0x00).copyInto(payload)
        biquad.forEachIndexed { i, b -> payload[7 + i] = b.toInt() and 0xFF }
        intArrayOf(f and 0xFF, f shr 8, q and 0xFF, q shr 8, g and 0xFF, g shr 8, w.typeCode, 0x00, slot, END)
            .copyInto(payload, 27)
        return report(*payload)
    }

    /** Global gain write `01 03 02 00 g`, g = round(dB) as int8. */
    fun preampWriteReport(db: Double): ByteArray = report(WRITE, CMD_GLOBAL_GAIN, 0x02, 0x00, toInt32(jsRound(db)) and 0xFF)

    /** devicePEQ commit after the filter writes - PERSISTS to flash. */
    val COMMIT_SEQUENCE: List<ByteArray> = listOf(
        report(WRITE, 0x05, END),
        report(WRITE, 0x17, END),
        report(WRITE, 0x0A, 0x04, 0x00, 0x00, 0xFF, 0xFF, END), // TEMP_WRITE
        report(WRITE, 0x01, 0x01, END),                         // FLASH_EQ enable + persist
    )
    private val COMMIT_DELAYS_MS = longArrayOf(20, 20, 50, 0)

    /** One report and the pause after it. */
    class Step(val report: ByteArray, val delayAfterMs: Long)

    /**
     * `write_state(...)` in research/protocol/write_test.py: every band with 20 ms after it, 100 ms more,
     * the preamp and 50 ms, then (when [commit]) the commit sequence with devicePEQ's delays; 200 ms at the end.
     */
    fun writeSequence(bands: List<BandWrite>, preampDb: Double, slot: Int, commit: Boolean): List<Step> {
        val steps = ArrayList<Step>()
        bands.forEachIndexed { i, b -> steps += Step(bandWriteReport(b, slot), if (i == bands.lastIndex) 120 else 20) }
        if (!commit) {
            steps += Step(preampWriteReport(preampDb), 50 + 200)
            return steps
        }
        steps += Step(preampWriteReport(preampDb), 50)
        COMMIT_SEQUENCE.forEachIndexed { i, r ->
            steps += Step(r, COMMIT_DELAYS_MS[i] + if (i == COMMIT_SEQUENCE.lastIndex) 200 else 0)
        }
        return steps
    }
}
